products/views.py

```diff
@@ -1,3 +1,73 @@
-from django.shortcuts import render
+import json
 
-# Create your views here.
+from django.http      import JsonResponse
+from django.views     import View
+from django.db.models import Q
+
+from datetime       import datetime
+from .models        import Category, Subcategory, Product
+from users.models   import User, Host
+
+class ProductView(View): 
+    def get(self, request):
+        try: 
+            category    = request.GET.get('category',None)
+            subcategory = request.GET.get('subcategory',None)
+            region      = request.GET.get('region',None)
+            group       = request.GET.get('group',None)
+            price       = request.GET.get('price',None)
+            offset      = int(request.GET.get("offset",0))
+            limit       = int(request.GET.get("limit",12))
+            offset      = offset * limit
+            limit       = offset + limit
+            
+            q_object = Q() 
+            
+            if category: 
+                q_object &= Q(subcategory__category=category)
+            if subcategory: 
+                q_object &= Q(subcategory_id=subcategory)
+            if region:
+                q_object &= Q(region=region)
+            if group:
+                q_object &= Q(is_group=group)
+
+            products = Product.objects.select_related('subcategory').filter(q_object)[offset:limit]
+            total_product_count = Product.objects.all().count()
+            
+            if price: 
+                products = Product.objects.select_related('subcategory').filter(q_object).order_by(price)[offset:limit]
```

Comment:
`select_related`로 subcategory를 join해서 가져오고 있는데, 아래에서 사용하지 않는 것 같아요. Django query loggin하는 방법을 찾아보시고, 쿼리를 최적화 해보시면 어떨까요?

```suggestion
                products = Product.objects.select_related('subcategory').filter(q_object).order_by(price)[offset:limit]
```

Reply:
select *
  from products
 inner join subcategories
    on products.subcategory_id = subcategories.id
 where subcategories.category_id = (select id from categories where id = 1);

Django query loggin 해보니까 이렇게 나오는것을 확인했습니다. 
여기서 더욱 최적화 해보려 했지만 쉽지 않네요 ㅜㅜㅜ

Reply:
위에 쿼리는 더이상 최적화 할게 없습니다! 아래 코드에서 for문이 실행될때 많은 쿼리가 발생한다는 얘기였습니다.

```python
results = [{
                'id'      : product.id,
                'title'   : product.title,
                'price'   : product.price,
                'bgimg'   : product.background_url,
                'userImg' : product.host.profile_url,
                'name'    : product.host.user.name,
                'nick'    : product.host.nickname
            } for product in products]
```

그래서 결론적으로 왜 `select_related('subcategory')`를 하시는지 궁굼합니다.

Reply:
제가 select_related를 제대로 이해하지 못하고 적용한것 같습니다 ㅜㅜ 
다시 공부한 뒤에 'subcategory'를 'host','host__user'로 변경해 보았습니다. 
results값에서는 product의 host와 host에 연결된 user만 가지고 오더라구요!

Reply:
👍 이제 제대로 잘 적용하신 것 같습니다! 4시에 공유되는 Django ORM 최적화 세션영상 보시면서 다시 개념 잡아보시면 좋을 것 같아요!